File: integration/testdata/kotlin/database/Echo.kt
package ftl.echo

import xyz.block.ftl.Context
import xyz.block.ftl.Verb
import xyz.block.ftl.Database

data class InsertRequest(val data: String)
typealias InsertResponse = Unit

val db = Database("testdb")

class Echo {

  @Verb
  fun insert(context: Context, req: InsertRequest): InsertResponse {
    persistRequest(req)
    return InsertResponse
  }

  fun persistRequest(req: InsertRequest) {
    db.conn {
      it.prepareStatement(
        """
        CREATE TABLE IF NOT EXISTS requests
        (
          data TEXT,
          created_at TIMESTAMPTZ NOT NULL DEFAULT (NOW() AT TIME ZONE 'utc'),
          updated_at TIMESTAMPTZ NOT NULL DEFAULT (NOW() AT TIME ZONE 'utc')
       );
       """
      ).execute()
      it.prepareStatement("INSERT INTO requests (data) VALUES ('${req.data}');")
        .execute()
    }
  }
}

File: integration/testdata/go/database/echo.go
package echo

import (
	"context"

	"github.com/TBD54566975/ftl/go-runtime/ftl" // Import the FTL SDK.
)

var db = ftl.PostgresDatabase("testdb")

type InsertRequest struct {
	Data string
}

type InsertResponse struct{}

//ftl:export
func Insert(ctx context.Context, req InsertRequest) (InsertResponse, error) {
	err := persistRequest(ctx, req)
	if err != nil {
		return InsertResponse{}, err
	}

	return InsertResponse{}, nil
}

func persistRequest(ctx context.Context, req InsertRequest) error {
	_, err := db.Get(ctx).Exec(`CREATE TABLE IF NOT EXISTS requests
	       (
	         data TEXT,
	         created_at TIMESTAMPTZ NOT NULL DEFAULT (NOW() AT TIME ZONE 'utc'),
	         updated_at TIMESTAMPTZ NOT NULL DEFAULT (NOW() AT TIME ZONE 'utc')
	      );`)
	if err != nil {
		return err
	}
	_, err = db.Get(ctx).Exec("INSERT INTO requests (data) VALUES ($1);", req.Data)
	if err != nil {
		return err
	}
	return nil
}
